python/ray/tune/variant_generator.py

```diff
@@ -0,0 +1,275 @@
+                checkpoint_freq=args.checkpoint_freq,
+                restore_path=spec.get("restore"),
+                upload_dir=args.upload_dir)
+
+
+def generate_variants(unresolved_spec):
+    """Generates variants from a spec (dict) with unresolved values.
+
+    There are two types of unresolved values:
+
+        Grid search: These define a grid search over values. For example, the
+        following grid search values in a spec will produce six distinct
+        variants in combination:
+
+            "activation": grid_search(["relu", "tanh"])
+            "learning_rate": grid_search([1e-3, 1e-4, 1e-5])
+
+        Lambda functions: These are evaluated to produce a concrete value, and
+        can express dependencies or conditional distributions between values.
+        They can also be used to express random search (e.g., by calling
+        into the `random` or `np` module).
+
+            "cpu": lambda spec: spec.config.num_workers
+            "batch_size": lambda spec: random.uniform(1, 1000)
+
+        It is also possible to nest the two, e.g. have a lambda function
+        return a grid search or vice versa, as long as there are no cyclic
+        dependencies between unresolved values.
+
+    Finally, to support defining specs in plain JSON / YAML, grid search
+    and lambda functions can also be defined alternatively as follows:
+
+        "activation": {"grid_search": ["relu", "tanh"]}
+        "cpu": {"eval": "spec.config.num_workers"}
+    """
+    for resolved_vars, spec in _generate_variants(unresolved_spec):
+        assert not _unresolved_values(spec)
+        yield _format_vars(resolved_vars), spec
+
+
+def grid_search(values):
+    """Convenience method for specifying grid search over a value."""
+
+    return {"grid_search": values}
+
+
+_STANDARD_IMPORTS = {
+    "random": random,
+    "np": numpy,
+}
+
+_MAX_RESOLUTION_PASSES = 20
+
+
+def _format_vars(resolved_vars):
+    out = []
+    for path, value in sorted(resolved_vars.items()):
+        if path[0] in ["alg", "env", "resources"]:
+            continue  # these settings aren't usually search parameters
+        pieces = []
+        last_string = True
+        for k in path[::-1]:
+            if isinstance(k, int):
+                pieces.append(str(k))
+            elif last_string:
+                last_string = False
+                pieces.append(k)
+        pieces.reverse()
+        out.append("_".join(pieces) + "=" + str(value))
+    return ",".join(out)
+
+
+def _generate_variants(spec):
+    spec = copy.deepcopy(spec)
+    unresolved = _unresolved_values(spec)
+    if not unresolved:
+        yield {}, spec
+        return
+
+    grid_vars = []
+    lambda_vars = []
+    for path, value in unresolved.items():
+        if isinstance(value, types.FunctionType):
+            lambda_vars.append((path, value))
+        else:
+            grid_vars.append((path, value))
+    grid_vars.sort()
+
+    grid_search = _grid_search_generator(spec, grid_vars)
```

Comment:
Would be nice to document the below logic perhaps just in the top of the function